feat: Add getSupportedLenses() stub and lensType controller parameter

hopingsteam opened the pull request.

hopingsteam committed:
Add support for camera lens types in MobileScannerController

hopingsteam committed:
Refactored getSupportedLenses to return empty list for no available lenses and updated documentation

hopingsteam committed:
Applied code formatting & removed redundant test

hopingsteam commented on test/mobile_scanner_controller_lens_type_test.dart:
```diff
@@ -0,0 +1,133 @@
+      expect(controller.lensType, CameraLensType.zoom);
+
+      await controller.dispose();
+    });
```
Sovled

navaronbracke commented on test/get_supported_lenses_test.dart:
```diff
@@ -0,0 +1,310 @@
+
+  group('getSupportedLenses() tests', () {
+    late MethodChannelMobileScanner platform;
+    late List<MethodCall> methodCalls;
```
I see that not every test uses this variable in an expect. Can we provide this variable only scoped to the tests that actually need it? The tests that do not need it will also not have to call `add()`

navaronbracke commented on test/get_supported_lenses_test.dart:
```diff
@@ -0,0 +1,310 @@
+              methodCalls.add(methodCall);
+
+              switch (methodCall.method) {
+                case 'getSupportedLenses':
```
I think we should provide the names of the supported methods as constants in the plugin.
Filed https://github.com/juliansteenbakker/mobile_scanner/issues/1606 to track this.

We can follow up with that after work on this feature is done.

navaronbracke commented on test/get_supported_lenses_test.dart:
```diff
@@ -0,0 +1,310 @@
+      methodCalls = <MethodCall>[];
+
+      TestDefaultBinaryMessengerBinding.instance.defaultBinaryMessenger
+          .setMockMethodCallHandler(
```
Doing `setMockMethodCallHandler()` in `setUp()` feels a bit redundant, since the tests will do their own version of `setMockMethodCallHandler()` anyway, which leads to a bit of duplication.

Can we remove this invocation of setMockMethodCallHandler and work with tear downs in the tests themselves? I think you can use a separate utility method in the tests for the tear down, since `platform` is set during the set up of individual tests:

```dart
void resetMethodCallHandler() {
  TestDefaultBinaryMessengerBinding.instance.defaultBinaryMessenger
          .setMockMethodCallHandler(platform.methodChannel, null);
}
```

navaronbracke commented on test/get_supported_lenses_test.dart:
```diff
@@ -0,0 +1,310 @@
+              methodCalls.add(methodCall);
+
+              if (methodCall.method == 'getSupportedLenses') {
+                return [0, 1, 2]; // normal, wide, zoom
```
Can we reuse the `rawValue` of the enum?

navaronbracke commented on test/get_supported_lenses_test.dart:
```diff
@@ -0,0 +1,310 @@
+      expect(lenses.first, CameraLensType.any);
+    });
+
+    test('handles unknown integer values (maps to any)', () async {
```
What with for example `[0, -99, 1]` ?

navaronbracke commented on test/start_options_test.dart:
```diff
@@ -0,0 +1,241 @@
+
+      final map = options.toMap();
+
+      expect(map['facing'], CameraFacing.back.rawValue);
```
Can we build a `Map<String, Object?>` with all the expected values and do one `expect()` call for the entire thing?

navaronbracke commented on test/start_options_test.dart:
```diff
@@ -0,0 +1,241 @@
+    });
+
+    test('toMap correctly maps all lens types', () {
+      const lensTypes = [
```
Can we use a for loop on `CameraLensType.values` , so that we do not miss any?

navaronbracke commented on test/start_options_test.dart:
```diff
@@ -0,0 +1,241 @@
+    });
+
+    test('toMap handles all camera facing directions', () {
+      const facingDirections = [
```
Same remark about using `.values`

navaronbracke commented on test/start_options_test.dart:
```diff
@@ -0,0 +1,241 @@
+
+    test('toMap handles boolean flags correctly', () {
+      const testCases = [
+        (returnImage: true, torch: true, invertImage: true, autoZoom: true),
```
We do not test all variants of each flag combination? (not sure if it is important to do so?)

hopingsteam replied:
The flags are independent - they don't interact with each other in `toMap()`. It just passes each value through, so testing all **16 permutations** would essentially be testing the same simple assignment repeatedly. The three cases cover: all true, all false, and a mixed scenario, which should be enough to catch any mapping issues.

navaronbracke commented on test/mobile_scanner_controller_lens_type_test.dart:
```diff
@@ -0,0 +1,119 @@
+
+      expect(controller.lensType, CameraLensType.any);
+
+      await controller.dispose();
```
I do not think we need to call dispose (since we never call `start()` ?) ? This test only calls the constructor of the controller, to verify instance properties, and does not "start" a (fake) camera.

The same is true for some of the other tests in this file.

navaronbracke commented on test/mobile_scanner_controller_lens_type_test.dart:
```diff
@@ -0,0 +1,119 @@
+      await controller.dispose();
+    });
+
+    test('different controllers can have different lens types', () async {
```
While the test itself somewhat makes sense, the plugin currently does not support running multiple controllers side by side. This is because it is unclear which controller should get ownership of the single logical camera driver (regardless of lenses)

I would omit this test.

navaronbracke commented on test/mobile_scanner_controller_lens_type_test.dart:
```diff
@@ -0,0 +1,119 @@
+      await zoomController.dispose();
+    });
+
+    test('controller with all lens types can be created', () async {
```
Isn't this duplicating the individual tests for the single lens type values?

I do prefer keeping this single test, since it covers all lens types.

navaronbracke commented on test/mobile_scanner_controller_lens_type_test.dart:
```diff
@@ -0,0 +1,119 @@
+      }
+    });
+
+    test('lens type property is read-only and immutable', () async {
```
This test makes no sense? You don't actually do `controller.lensType = someOtherType;` in the test.
And because it is a `final` property, you won't be able to anyway.

I would omit this test.

navaronbracke reviewed (changes requested):
I went ahead and gave it a review. Overall it is nice to see some extra test coverage!
I do have some feedback, though.

hopingsteam committed:
Fixed PR comments

navaronbracke commented on test/get_supported_lenses_test.dart:
```diff
@@ -0,0 +1,330 @@
+      expect(lenses.first, CameraLensType.any);
+    });
+
+    test('filters unknown values while keeping valid ones', () async {
```
Is this a typo? If I understand correctly, invalid values map to `CameraLensType.any`, so it does not "filter" anything. It maps invalid values to the "any" value. So maybe the test name should be:

`"maps unknown values to 'any' lens type"`

Also, what happens if we have `0, -99, 1, -10`, do we get multiple "any" lens types?
I wonder if the platform interface should return a `Set` and not a `List` then?

hopingsteam replied:
The test name was misleading - it's mapping unknown values to any, not filtering them out. I'll rename it.

navaronbracke replied:
Still, would we then get "any" multiple times? It would only make sense if it was included once (if present at all)?

hopingsteam replied:
I've pushed your suggestion!

navaronbracke reviewed (changes requested):
Almost there! One small remark still.

hopingsteam committed:
Renamed test case

hopingsteam committed:
Merge branch 'develop' into stub_implementation

hopingsteam committed:
Refactor to return a set + updated the tests for validating the implementation

hopingsteam commented:
I've changed the return type from `List` to `Set` so duplicates are automatically handled. Now if the platform returns something like `[normal, -99, wide, -10, 99]`, we get `{normal, wide, any}` instead of having `any` appear 3 times.

Also merged the `@visibleForTesting` constants PR and updated the tests to use them.

navaronbracke commented on lib/src/mobile_scanner_controller.dart:
```diff
@@ -576,6 +590,42 @@ class MobileScannerController extends ValueNotifier<MobileScannerState> {
+  ///
+  /// Returns a set of [CameraLensType] values that are available on the
+  /// device. This can be used to determine which lens types can be used
+  /// with the scanner. Using a set ensures no duplicate lens types are
```
The line `Using a set ensures no duplicate lens types are returned.` is redundant, since that is what a Set datatype does anyway. (it does not allow multiple equal objects)

navaronbracke commented on lib/src/mobile_scanner_platform_interface.dart:
```diff
@@ -117,6 +118,19 @@ abstract class MobileScannerPlatform extends PlatformInterface {
+  /// Get the set of supported camera lens types for the current device.
+  ///
+  /// Returns a set of [CameraLensType] values that are available on the
+  /// device. Using a set ensures no duplicate lens types are returned.
```
Same here for the line `Using a set ensures no duplicate lens types are returned.`

navaronbracke reviewed (changes requested):
One small remark about doc comments, but otherwise this is ready to merge.

hopingsteam committed:
Removed redundant comments

hopingsteam commented:
I added those comments because I was changing from `List` to `Set` and wanted to explain the reasoning behind the change. But as you've said - explaining what a Set does is redundant like explaining that water is wet
